Fix forkr - Handle rng in Trace

FredericWantiez opened the pull request:
Still very early, just opening it to track changes/ideas.

devmotion commented:
Maybe it would be good to merge https://github.com/TuringLang/AdvancedPS.jl/pull/20 first?

FredericWantiez commented:
Agree, we need to build on #20 for this fix.

yebai commented:
> Agree, we need to build on #20 for this fix.

I've merged #20.

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,31 @@
+    return res
+end
+
+inc_count!(rng::TracedRNG) = inc_count!(rng, 1)
```
To me it still seems that [counter-based PRNGs](https://en.wikipedia.org/wiki/Counter-based_random_number_generator_(CBRNG)) might be better here: they would allow to (re-)generate the n-th random number without having to generate the first n-1 random numbers.

devmotion replied:
I think the implementation is actually in https://github.com/sunoru/Random123.jl.

devmotion replied:
An example on the GPU: https://juliafolds.github.io/FoldsCUDA.jl/dev/examples/monte_carlo_pi/

yebai replied:
I like the idea, however, I would suggest that we keep this as future work. For this PR, we probably want to pass `rng` around all necessary places first, so that we can fix `forkr` next.

devmotion replied:
:+1: 

I just noticed that DiffEqNoiseProcesses also uses splittable counter-based RNGs: https://github.com/SciML/DiffEqNoiseProcess.jl/blob/master/src/noise_interfaces/virtual_brownian_tree_interface.jl and https://diffeq.sciml.ai/stable/features/noise_process/#VirtualBrownianTree

devmotion commented on src/smc.jl:
```diff
@@ -38,10 +38,10 @@ function AbstractMCMC.sample(
 
     # Create a set of particles.
-    particles = ParticleContainer([Trace(model) for _ in 1:sampler.nparticles])
+    particles = ParticleContainer([Trace(model, rng) for _ in 1:sampler.nparticles])
```
I think it is crucial to not use the user-provided RNG directly if one wants to keep track of the number of generated random numbers in order to achieve reproducibility: if the user updates the state of the RNG outside of the trace (e.g., by generating other random numbers, in particular if it is just the default RNG) the sequence of random numbers in the trace can not be reproduced anymore.

FredericWantiez replied:
Good point, maybe we can make a copy of the rng and pass it to Trace ?

FredericWantiez commented:
@yebai I think there's an issue with using fork and passing the RNG to Trace. When we fork the particles during the resampling step, the RNG is set to the RNG we used in the original Trace definition, all children particles will use it and we won't be able to reconstruct the stream.

Also, on the side, when do we populate the reference particle when we resample in PG ?

yebai reviewed (changes requested):
We'll need a mechanism to split the random stream during the resampling step. The splitting operation 

1. allows the children particles to share the random history before the splitting point, 
2. make the random streams produce independent random numbers after the splitting point.

> Also, on the side, when do we populate the reference particle when we resample in PG ?

I'm not sure I understand the question. The reference particle is incrementally replayed during the forward propagation of a conditional SMC step. If the reference particle has multiple children at a resampling step, one child will inherit the random stream and keep replaying, the other children will split the random stream and start producing new random numbers. 

Also, happy to discuss more when we meet.

devmotion commented:
> We'll need a mechanism to split the random stream during the resampling step. The splitting operation
> 
>     1. allows the children particles to share the random history before the splitting point,
> 
>     2. make the random streams produce independent random numbers after the splitting point.

Isn't this exactly what splittable RNGs are designed for (e.g. https://github.com/google/jax/blob/master/design_notes/prng.md#design, https://dl.acm.org/doi/abs/10.1145/2660193.2660195, and https://publications.lib.chalmers.se/records/fulltext/183348/local_183348.pdf)? The last reference also discusses why naive approaches for splitting linear PRNGs fail or make it hard to assess the quality of the returned random numbers (I am not sufficiently familiar with this area to determine if these all are valid concerns).

yebai commented:
yes, we are planning to use splittable RNGs.

FredericWantiez commented:
Looked at the random module in tensorflow recently, they also use counter-based rng and the split api: https://github.com/tensorflow/tensorflow/blob/v2.5.0/tensorflow/python/ops/stateful_random_ops.py#L933-L981

We could do something similar:
```julia
mutable struct TracedRNG{R,T<:Random123.AbstractR123{R}}
    count::Int
    rng::T
    keys::Array{R}
end


TracedRNG(r::Random123.AbstractR123) = TracedRNG(0, r, [r.key])


function split(r::TracedRNG, n::Integer)
    rngs = map(1:n) do i
        new_seed = hash(r.rng.key, r.rng.ctr1+i)
        new_rng = TracedRNG(r.count, Philox2x(new_seed), [r.keys..., new_seed])
        return new_rng
    end
end

rng = TracedRNG(Philox2x())
a, b = split(rng, 2)  
```                                                                                                                                                                                                                                                                                                                            
To replay the random numbers we should be able to cycle through the keys and reset the counters

FredericWantiez commented:
@yebai I simplified the API slightly, we don't need to track both the model counter and the rng counters. We can use the model counter if every particle has an unique key/id. Fixed the naming and the resampling step that was wrong. Also added a few tests. I can add some more, don't think we catch all the edge cases.

yebai reviewed (changes requested):
Thanks @FredericWantiez - excellent work. I have gone through the `rng.jl` file and made some comments along the process. I'll take a look at other parts slightly later. 

In particular,  I find the function pair `split` and `update_rng!` is slightly confusing while reading the code. Given that the `split` and `update_rng!` functions are always used in pairs, I wonder whether we can consider merging these functions into `split!(rng::TracedRNG)::TracedRNG` for better clarity. This `split!` function should always split a given RNG once, mutate it in place and return the mutated RNG. If we need to split N-times, we simply call `split` N-times repetitively. Below is an example implementation of the `split!` function: 


```julia
function split!(rng::TracedRNG{T}) where {T}
    key = hash(r.rng.key, convert(UInt, r.rng.ctr1)
    seed!(rng, key)
    return set_counter!(rng.rng, rng.count + 1) # rng.count + 1 is used instead of rng.count
end
```

In order to ensure that calling `split!` repetitively with the same RNG will return new RNGs with different keys, we increase the internal counter by one every time a split is performed.  This way, the next time when `split!` is called on the same RNG, it will get a different key because `r.rng.ctr1` will be different. 

A side question: do we actually need to set `rng.rng.ctr1` to `rng.count` when performing a split? If so, is there a reference on this?

FredericWantiez commented:
@devmotion @yebai Fixed the last few thing we discussed last week, let me know what you think.

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+using Random123
+using Random
+using Distributions
```
Could these be moved to src/AdvancedPS.jl?

cpfiffer replied:
Yeah, I'd prefer to see all the imports in `src/AdvancedPS.jl`.

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+using Random
+using Distributions
+
+import Base.rand
+import Random.seed!
+import Random123: set_counter!
```
IMO it is safer and more readable to not import the extended functions but to define the new methods with
```julia
Base.rand(...) = ...
```
etc.

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+import Random123: set_counter!
+
+# Default RNG type for when nothing is specified
+_BASE_RNG = Philox2x
```
```suggestion
const _BASE_RNG = Philox2x
```

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+
+Wrapped random number generator from Random123 to keep track of random streams during model evaluation
+"""
+mutable struct TracedRNG{T} <:
+               Random.AbstractRNG where {T<:(Random123.AbstractR123{R} where {R})}
```
```suggestion
mutable struct TracedRNG{T<:Random123.AbstractR123} <: Random.AbstractRNG
```

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+    "Inner RNG"
+    rng::T
+    "Array of keys"
+    keys
```
Make this concretely typed as well?

FredericWantiez replied:
Maybe something like this: 
```julia
mutable struct TracedRNG{R, T<:Random123.AbstractR123{R}} <: Random.AbstractRNG
    "Model step counter"
    count::Int
    "Inner RNG"
    rng::T
    "Array of keys"
    keys::Array{R}
end
```

devmotion replied:
`Array{R}` is not a concrete type but e.g. `Vector{R}`, `Matrix{R}` or `Array{R,N}` (with additional type parameter `N`) would be.

FredericWantiez replied:
My bad, we need `Array{R,N}`. I'll change the types accordingly.

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+    keys
+end
+
+"""
+    TracedRNG(r::Random123.AbstractR123)
+
+Initialize TracedRNG with r as the inner RNG 
+"""
+function TracedRNG(r::Random123.AbstractR123)
+    set_counter!(r, 0)
+    return TracedRNG(1, r, typeof(r.key)[])
+end
+
+"""
+    TracedRNG()
+
+Create a default TracedRNG
+"""
+function TracedRNG()
+    r = _BASE_RNG()
+    return TracedRNG(r)
+end
```
A bit simpler:
```suggestion
"""
    TracedRNG(r::Random123.AbstractR123=AdvancedPS._BASE_RNG())

Create a `TracedRNG` with `r` as the inner RNG. 
"""
function TracedRNG(r::Random123.AbstractR123=_BASE_RNG())
    set_counter!(r, 0)
    return TracedRNG(1, r, typeof(r.key)[])
end
```

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+end
+
+# Connect to the Random API
+Random.rng_native_52(rng::TracedRNG{U}) where {U} = Random.rng_native_52(rng.rng)
+Base.rand(rng::TracedRNG{U}, ::Type{T}) where {U,T} = Base.rand(rng.rng, T)
```
The type parameters are not needed:
```suggestion
Random.rng_native_52(rng::TracedRNG) = Random.rng_native_52(rng.rng)
Base.rand(rng::TracedRNG, ::Type{T}) where {T} = rand(rng.rng, T)
```

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+
+Split key into n new keys
+"""
+function split(key::Integer, n::Integer=1) where {T}
```
```suggestion
function split(key::Integer, n::Integer=1)
```

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+Split key into n new keys
+"""
+function split(key::Integer, n::Integer=1) where {T}
+    return map(i -> hash(key, convert(UInt, i)), 1:n)
```
A bit simpler, I think:
```suggestion
    return [hash(key, i) for i in UInt(1):UInt(n)]
```

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+"""
+    split(key::Integer, n::Integer)
+
+Split key into n new keys
```
```suggestion
Split `key` into `n` new keys.
```

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+Base.rand(rng::TracedRNG{U}, ::Type{T}) where {U,T} = Base.rand(rng.rng, T)
+
+"""
+    split(key::Integer, n::Integer)
```
```suggestion
    split(key::Integer, n::Integer=1)
```

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+
+Load state from current model iteration. Random streams are now replayed
+"""
+function load_state!(rng::TracedRNG{T}) where {T}
```
```suggestion
function load_state!(rng::TracedRNG)
```

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+
+Set key and counter of inner RNG to key and the running model step
+"""
+function seed!(rng::TracedRNG{T}, key) where {T}
```
```suggestion
function seed!(rng::TracedRNG, key)
```

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+
+Track current key of the inner RNG
+"""
+function save_state!(r::TracedRNG{T}) where {T}
```
```suggestion
function save_state!(r::TracedRNG)
```

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+    return push!(r.keys, r.rng.key)
+end
+
+Base.copy(r::TracedRNG{T}) where {T} = TracedRNG(r.count, copy(r.rng), copy(r.keys))
```
```suggestion
Base.copy(r::TracedRNG) = TracedRNG(r.count, copy(r.rng), copy(r.keys))
```
I am also worried that `copy(r.rng)` is not sufficient to ensure that the copy is decoupled from the original RNG. Maybe it would be safer (albeit more expensive) to use `deepcopy`?

devmotion commented on src/rng.jl:
```diff
@@ -0,0 +1,104 @@
+"""
+    inc_counter!(r::TracedRNG, n::Integer=1)
+
+Increase the model step counter by n
```
```suggestion
Increase the model step counter by `n`.
```

devmotion commented on test/smc.jl:
```diff
@@ -149,6 +150,34 @@
     end
+
+    @testset "Replay reference" begin
+        mutable struct Model <: AbstractMCMC.AbstractModel
+            a::Float64
+            b::Float64
+
+            Model() = new()
+        end
```
IIRC one has to define structs at the top level.

FredericWantiez replied:
Seems like we define the structs locally in most of `test/smc.jl` but I can move it at the top module.
https://github.com/TuringLang/AdvancedPS.jl/blob/6979bbef862589a2dbe4cad062911409c733333e/test/smc.jl#L66-L98

cpfiffer commented on src/container.jl:
```diff
@@ -48,16 +60,18 @@ end
     ctask = let f = trace.ctask.task.code
         Libtask.CTask() do
-            res = f()
+            res = f()(trace.rng)
```
Isn't `f` a function generator, here?

cpfiffer reviewed (commented):
I think David's already added most of the things I would add -- in general this seems to be fine with a few minor changes needed.

yebai commented on src/rng.jl:
```diff
@@ -0,0 +1,84 @@
+Split `key` into `n` new keys
+"""
+function split(key::Integer, n::Integer=1)
+    return [hash(key, i) for i in UInt(1):UInt(n)]
```
```suggestion
    T = typeof(key) # Make sure the type of `key` is consistent on W32 and W64 systems. 
    return T[hash(key, i) for i in UInt(1):UInt(n)]
```

yebai commented:
Excellent work - many thanks @FredericWantiez @devmotion and @cpfiffer! 

Perhaps we can wait for a few days to see whether we miss anything, then make a new release.

devmotion commented on src/container.jl:
```diff
@@ -1,31 +1,43 @@
-struct Trace{F}
+struct Trace{F,U,N,V<:Random123.AbstractR123{U}}
```
Do we dispatch on `U`, `N` or `V` somewhere? Otherwise (and probably even in this case) this could be simplified to
```julia
struct Trace{F,R<:TracedRNG}
    f::F
    ctask::Libtask.CTask
    rng::R
end
```

devmotion commented on src/container.jl:
```diff
@@ -81,15 +95,21 @@ Data structure for particle filters
 - consume(pc::ParticleContainer): return incremental likelihood
 """
-mutable struct ParticleContainer{T<:Particle}
+mutable struct ParticleContainer{T<:Particle,U,N,V<:Random123.AbstractR123{U}}
```
Same here, I guess we could use
```julia
mutable struct ParticleContainer{T<:Particle,R<:TracedRNG}
    vals::Vector{T}
    logWs::Vector{Float64}
    rng::R
end
```
?

devmotion commented on src/container.jl:
```diff
@@ -170,6 +193,22 @@ function effectiveSampleSize(pc::ParticleContainer)
+        k = split(pi.rng.rng.key)
+        Random.seed!(pi.rng, k[1])
+    end
+end
```
I guess we could add an empty `return` or `nothing` at the end to avoid that the last seed or rng is returned.

devmotion commented on test/container.jl:
```diff
@@ -11,7 +11,7 @@
         function fpc(logp)
             f = let logp = logp
-                () -> begin
+                (rng) -> begin
```
The brackets could be omitted.
